src/Lang/Surface.ml

```diff
@@ -87,21 +82,55 @@ and type_expr_data =
   | TVar of tvar path
     (** A (non-unification) type variable *)
 
+  | TEffect of type_expr list
+    (** Effect: list of simple effects *)
+
   | TPureArrow of scheme_expr * type_expr
     (** Pure function: a function without effects, that always terminates *)
 
   | TArrow of scheme_expr * type_expr * type_expr
     (** Effectful function: the last parameter is an effect *)
 
-  | TEffect of type_expr list * type_expr option
-    (** Effect: list of simple effect optionally closed by another effect *)
+  | THandler of (** First-class handler *)
+    { effect   : tvar;
+        (** Effect variable bound by this handler (it bound in [cap_type],
+          [in_type], and [in_eff] *)
+
+      cap_type : type_expr;
+        (** Type of a capability provided by this handler *)
+
+      in_type  : type_expr;
+        (** Inner type of a handler: a type of expression that can be run
+          inside this handler *)
+
+      in_eff   : type_expr;
+        (** Inner effect of a handler *)
+
+      out_type : type_expr;
+        (** Outer type of a handler: a type of the whole handler expression
+        *)
+
+      out_eff  : type_expr
+        (** Outer effect of a handler *)
+    }
+
+  | TE_Label of (** First-class label *)
+    { effect    : type_expr;
+        (** Effect of this label *)
+
+      delim_tp  : type_expr;
+        (** Type of the delimiter *)
+
+      delim_eff : type_expr
+        (** Effect of the delimiter *)
+    }
 
   | TApp of type_expr * type_expr
     (** Type application *)
 
 (** Type-scheme expressions *)
 and scheme_expr = {
-  sch_pos : Position.t;
+  sch_pos   : Position.t;
     (** Location of the scheme expression *)
 
   sch_targs : named_type_arg list;
```

Comment:
I think forcing named types to come before named value parameters in schemes and constructor declarations will seem inconsistent now that this restriction is lifted in patterns.

For example, I believe the following uses of similar-looking syntax should have similar behavior regarding binding visibility, but I'm not sure they will if Surface is defined like this.
```fram
data T = ...

data U = C of {x : T, T} # Same as {T, x : T}?

# Referring to the previously defined T for x instead?
let C {x : T, T} = ...
```